core/src/main/java/org/mule/runtime/core/internal/execution/FlowProcessMediator.java

```diff
@@ -515,13 +517,23 @@ private void onMessageReceived(CoreEvent event, Pipeline flowConstruct, DefaultF
         .fireNotification(notificationFunction.apply(event, flowConstruct.getSource())));
   }
 
+  private Optional<String> evaluateCorrelationIdExpressionGenerator() {
+    return muleContext.getConfiguration().getDefaultCorrelationIdGenerator().map(generator -> generator.generateCorrelationId());
```

Comment:
obtain the correlationIdGenerator in the constructor instead of resolving it here on every call. Also, resolve the optional there and avoid using `map` here on every call.